syntax = "proto3";

import "lte/protos/oai/std_3gpp_types.proto";
import "lte/protos/oai/common_types.proto";
import "lte/protos/oai/nas_state.proto";
import "lte/protos/oai/spgw_state.proto";
import "google/protobuf/timestamp.proto";

package magma.lte.oai;
option go_package = "magma/lte/cloud/go/protos/oai";

// sgs_context_t
message SgsContext {
  // TODO
}

/*****************************************************************/
/*                     messages for MME                          */
/*****************************************************************/
// bearer_context_t
message BearerContext {
  uint32 ebi = 1; // ebi_t
  uint32 transaction_identifier = 2; // proc_tid_t
  Fteid s_gw_fteid_s1u = 3; // fteid_t
  Fteid p_gw_fteid_s5_s8_up = 4; // fteid_t
  uint32 qci = 5; // qci_t

  uint32 pdn_cx_id = 6; // pdn_cid_t
  EsmEbrContext esm_ebr_context = 8;
  Fteid enb_fteid_s1u = 9;

  uint32 priority_level = 10; // priority_level_t
  uint32 preemption_vulnerability = 11;
  uint32 preemption_capability = 12;
  TrafficFlowTemplate saved_tft = 13;
  BearerQos saved_qos = 14;
  enum Validity {
    INVALID = 0;
    VALID = 1;
  }
  Validity validity = 15;
}

// esm_pdn_t
message EsmPdn {
  uint32 pti = 1;
  bool is_emergency = 2;
  uint32 ambr = 3;
  uint32 addr_realloc = 4;
  uint32 n_bearers = 5;
  uint32 pt_state = 6; // esm_pt_state_e enum
}

// pdn_context_t
message PdnContext {
  uint32 context_identifier = 1; // context_identifier_t
  string apn_in_use = 2;
  string apn_subscribed = 3;
  uint32 pdn_type = 4; // pdn_type_t
  bytes paa = 5; // paa_t
  string apn_oi_replacement = 6;
  string p_gw_address_s5_s8_cp = 7; // ip_address_t
  uint32 p_gw_teid_s5_s8_cp = 8; // teid_t
  EpsSubscribedQosProfile default_bearer_qos_profile = 9;
  Ambr subscribed_apn_ambr = 10;
  Ambr p_gw_apn_ambr = 11;
  uint32 default_ebi = 12; // ebi_t
  repeated uint32 bearer_contexts = 13;
  string s_gw_address_s11_s4 = 14; // ip_address_t
  uint32 s_gw_teid_s11_s4 = 15;
  EsmPdn esm_data = 16;
  bool is_active = 17;
  ProtocolConfigurationOptions pco = 18;
}

message Regional_subscription {
  string zone_code = 1;
}

// ue_mm_context_t
message UeContext {
  string msisdn = 1;
  uint32 rel_cause = 2; //enum s1cause
  uint32 mm_state = 3; // mm_state_t
  uint32 ecm_state = 4; // ecm_state_t

  EmmContext emm_context = 5;
  uint32 sctp_assoc_id_key = 6;

  uint32 enb_ue_s1ap_id = 7;
  uint64 enb_s1ap_id_key = 8;
  uint32 mme_ue_s1ap_id = 9;
  uint32 attach_type = 10;
  uint32 sgs_detach_type = 11;

  // eNB information
  Ecgi e_utran_cgi = 20;
  uint64 cell_age = 21;

  // location information
  bytes lai = 25; // lai_t

  // PDN information
  ApnConfigProfile apn_config = 30;
  uint32 sub_status = 31; // subscriber_status_t
  uint32 subscriber_status = 32; // subscriber_status_t
  uint32 network_access_mode = 33; // network_access_mode_t
  uint32 access_restriction_data = 34; // ard_t
  string apn_oi_replacement = 35;
  uint32 mme_teid_s11 = 36; // teid_t
  Ambr subscribed_ue_ambr = 37; //ambr_t
  Ambr used_ue_ambr = 38; // ambr_t
  Ambr used_ambr = 39; //ambr_t
  uint32 nb_active_pdn_contexts = 40;
  repeated PdnContext pdn_contexts = 41; //pdn_context_t
  repeated BearerContext bearer_contexts = 42; // bearer_context_t
  bytes ue_radio_capability = 43;

  // Regional subscription
  uint32 num_reg_sub = 44;
  repeated Regional_subscription reg_sub = 45;

  // flags
  bool send_ue_purge_request = 60;
  bool hss_initiated_detach = 61;
  bool location_info_confirmed_in_hss = 62;
  bool ppf = 63;
  bool subscription_known = 64;
  bool path_switch_req = 65;

  // CSFB
  uint32 granted_service = 100; // granted_service_t
  uint32 cs_fallback_indicator = 101;
  SgsContext sgs_context = 102;

  Timer mobile_reachability_timer = 200;
  Timer implicit_detach_timer = 201;
  Timer initial_context_setup_rsp_timer =202;
  Timer ue_context_modification_timer = 203;
  Timer paging_response_timer = 204;
  uint32 rau_tau_timer = 205;
  Timer ulr_response_timer = 206;
  google.protobuf.Timestamp time_mobile_reachability_timer_started = 207;
  google.protobuf.Timestamp time_implicit_detach_timer_started = 208;
  google.protobuf.Timestamp time_paging_response_timer_started = 209;
  uint32 paging_retx_count = 210;
  google.protobuf.Timestamp time_ics_rsp_timer_started = 211;
}

// mme_ue_context_t
message MmeUeContext {
  map<uint64, uint64> imsi_ue_id_htbl = 4;  // imsi -> mme_ue_id
  map<uint64, uint64> tun11_ue_id_htbl = 5; // tun_id -> mme_ue_id
  map<uint64, uint64> enb_ue_id_ue_id_htbl = 7; // enb_ue_id -> mme_ue_id
  map<string, uint64> guti_ue_id_htbl = 8; // guti -> mme_ue_id
}

// mme_app_desc_t
message MmeNasState {
  MmeUeContext mme_ue_contexts = 1; // mme_ue_context_t
  uint32 nb_enb_connected = 2 [deprecated=true];
  uint32 nb_ue_attached = 3;
  uint32 nb_ue_connected = 4;
  uint32 nb_default_eps_bearers = 5;
  uint32 nb_s1u_bearers = 6;
  uint64 statistic_timer_id = 7;
  uint32 nb_ue_managed = 8;
  uint32 nb_ue_idle = 9;
  uint32 nb_bearers_managed = 10;
  uint32 nb_ue_since_last_stat = 11; //TODO: remove
  uint32 nb_bearers_since_last_stat = 12; //TODO: remove
  uint32 mme_app_ue_s1ap_id_generator = 20;
}
